README.md
# Required modules
simply run

`sudo pip install -r ./requirements.txt`

or

`sudo pip3 install -r ./requirements.txt`

This program is py2 && py3 compatiable after 2018.03.23 modification(theorically...)

# How to use
`python2 run_me.py` or `python3 run_me.py`


# Connection
- Communicate with [ADS1299](http://www.ti.com/product/ADS1299) through SPI
interface on OrangePi pin `CS/PA13`, `CLK/PA14`, `MOSI/PA15` and `MISO/PA16`

- On shield Atmega328P(Arduino Uno/Nano) is connected to `UART2_TX/PA00` and
`UART2_RX/PA01`

- Control SSD1306 0.96' OLED screen by SPI or ILI9325 2.3' LCD screen through
on shield Arduino by serial

- Broadcast collected raw data from ADS1299 to wifi port 9999(default) by TCP
socket, there are two ways to grab data from network
    - Connect OrangePi to your LAN wifi network and check its IP address, you
    can login to router(e.g. TP-Link @ `192.168.1.1`) or use any other methods
    to get OrangePi IP address.
    - Set OrangePi as a wifi hotpot, connect your PC/laptop/phone to this
    network and this way OrangePi IP address will be `192.169.0.1`(usually)


## Get data from matlab
- Make sure orangepi and your PC are in same LAN
- Here's an example script to fetch data through socket in matlab

```Matlab
% socket server on orangepi default listen on port 9999.
client = tcpclient('192.168.10.10', 9999)

% 8-channel float32 data --> 8_ch * 32_bits / 8_bits = 32 bytes data
data = client.read(32)

% unpack bytes into float32(single)
data = typecast(data, 'single')

% here data is 1x8 vector
data
```


# Project structure
|    folder    |    description    |
| :----------- | :---------------- |
|     data     | save biosignal data with label in each subfolder named by username |
|    models    | save trained models with weight for each user |
|     src      | preprocessing algorithms, classifiers and frameworks |
|    test      | testing new ideas |
|    utils     | common functions, gym clients, data IO, etc. |
|   run_me.py  | bootloader script of recognition program |
|   run_me1.py | bootloader script of Screen GUI |


# Supported gyms
Currently we only support two environments
- plane-war-game: written by [@buaawyz](https://github.com/buaawyz),
[installation guide](https://github.com/hankso/gym_plane_python),
run `python main.py` first and then `from gyms import PlaneClient as Client`
- torcs-car-game: see more details at [gym_torcs](https://github.com/ugo-nama-kun/gym_torcs)


## EmBCI
EmBCI is Embedded Brain Computer Interface, a bio-signal acquisition and processing platform.

It's composed of a **Hardware**, a **Python library** and a **Linux Service Interface**.

## Hardware

## Python lib

## Service
EmBCI has a `Linux` service interface to work properly on embedded devices.

More details at [files/service](files/service).


## Install
### Debian & Ubuntu & Windows
`pip install embci` **(to be implemented, not working now)**

### Install from source
```bash
git clone git@github.com:hankso/EmBCI.git && cd EmBCI
sudo python -m pip install -r ./requirements.txt
python setup.py build && sudo python setup.py install
```

This program is py2 && py3 compatiable.

## Output Interface
### Get data from matlab
- Here we use Orange Pi + EmBCI Shield Rev.A7
- Make sure Orange Pi and your PC are in same LAN
- Here's an example script to fetch data through socket in `Matlab`

```Matlab
% socket server on orangepi default listen on port 9999.
client = tcpclient('192.168.10.10', 9999)

% 8-channel float32 data --> 8_ch * 32_bits / 8_bits = 32 bytes data
data = client.read(32)

% unpack bytes into float32(single)
data = typecast(data, 'single')

% here data is 1x8 vector
data
```

## Project structure
|    folder     |    description    |
| :------------ | :---------------- |
|    embci      | Preprocessing algorithms, classifiers, WebUI and frameworks |
|   files/avr   | On shield Atmega328P firmware |
|  files/esp32  | On shield ESP32 firmware and burning tools |
| files/layouts | Saved SPI-Screen GUI layouts in `python-pickle` format |
| files/service | Linux service files |


## TODOs
### Documents
- This page & README
- embci lib docs

### Hardware Design
- ESP32
    - Serial connection higher baudrate (like 921600)
    - Act as a portable WiFi card: WiFi-echo (through Serial/I2C)
    - Change command interface from SPI to Serial
- ADS1299 better de-noising
- PCB shielding cases

### Algorithums
- SSVEP & P300
- ERP/EDP
- Motor Imagery

### Application
- Parkinson DBS treatment recovery
- Online SSVEP mind-typing
- Sign language sEMG recognition


Changelog.md

## 0.2.3 [2019-10-22]
- [Speller] Save user's configurations in browser to recover the same layout at next visit. Add width & height attributes to each layout. Copy recognition result to clipboard by double clicking result field. Finish all TODOs.
- [DisplayWeb] Add a second rendering method "scroll whole frame". This can be enabled by one click of a toggle button. Optimize realtime render performance.
- [Recorder] Fix recorder command entrance multi-command bug. Provide CLI data stream recorder application at `python -m embci.apps.recorder` (like a LSL LabRecorder implemented in Python) which supports stream selection, recording pause & resume and data filename/username configuration.
- [Recorder] Add web-based recorder console at `http://${EMBCI_HOST}/app/recorder`. List and control all alive recorders in your browser.
- [Streaming] Add JSON-RPC interface to control ESP32 data stream and LSL outlet by Remote Procedure Call. JSON-RPC protocal support communication from C/Python/JavaScript/... See more at module doc `embci.apps.streaming.__doc__`.
- [WebUI] Add `webui_static_host` function to register multiple source directories for static file serving. It supports updating on exist routes.
- [File Manager] New feature: data files manager in subapp `embci.apps.baseserver`.
- [JSONRPC] History container informatic summary output. Allow CORS POST by setting `Access-Control-Allow-Origin` in response header.
- Enable subapp `embci.apps.system` to shutdown/reboot/update the EmBCI device. Use `embedded_only` decorator to avoid dangerous command on PC.

## 0.2.2 [2019-09-29]
- [logging] Use single character in logging level output. Support colorful output in shell/terminal.
- [IO] Add `save_chunk(append=False)` and `save_trial`.
- [Recorder] Subapp recorder well tested. Check `embci.apps.recorder.__doc__` for more detail about configuration and usage.

## 0.2.1 [2019-09-15]
- Add `from __future__ import absolute_import, division, print_function` to each module.
- Use ``//`` for floor division and ``/`` for float division (immigrating to py3).
- [WebUI] HTML snippets loader support keywords arguments by setting attributes like ``data-foo="bar"``. Add webui debug application runner and pick random port.
- [Drivers] Add package `embci.drivers`. Move `embci.utils.xxx_api` to `embci.drivers.xxx` as hardware APIs.
- [Utils] Add `get_free_port` to resolve an avaiable port number. Support event sending and receiving by EventIO system implemented at `embci.utils._event`
- [Speller] Subapp SSVEP mind-typing constructed. Will be tested soon.
- [Recorder] Add snippet `div#recorder` in `recorder-box.html`.

## 0.2.0 [2019-09-01]
- [Apps] Project `DBS` is removed from EmBCI and will be maintained by a business company.
- [Apps] Remove `DisplaySPI` and add subapp `DisplayWeb`.
- [Utils] Move `minimize` as standalone function. Logger stack info py 2 & 3 compatiable. Add `validate_filename` & `Namespace`.
- [Utils] Add `obfuscation` to distribute python code in dynamic linked library format.
- [JSONRPC] Add `NoExist` type and support for multiple instance registration.
- [WebUI] Add package `Numjs`. Rename `xxxDIR` to `DIR_xxx` and `__dir__` to `__basedir__`. Support refresh/rescan subapps at runtime.
- [WebUI] Support preventing subapps from loading by command line arguments `python -m embci.webui --exclude app1 app2`
- [Tools] Move `embci_tool` to `embci.apps.system`. Add template.py for python source files.
- [Setup.py] Load requirements.txt from absolute path.
- [configs] Load from configuration files when imported and export a dict at `embci.configs.settings`. Set `ENSURE_DIR_EXIST = True` in conf files if you want to create necessary directories.
- [io] Readers now get new valid name automatically by searching pidfiles. Add `hook_before` and `hook_after`.

## 0.1.6 [2019-07-04]
- [WebUI] Add supports for HTML snippets & navbar.html & bottom.html.
- [WebUI] Resolve battery level from ESP32 and render it to navbar.
- [ESP32] Support set/get Channel/DataSource/Gain/Impedance/SampleRate/Bias.
- [ESP32] Add second order IIR realtime filter and use looply-level-detection instead of interrupt.
- [Apps] Add subapp `auth`.
- [DBS] Hide DBS.utils source code with `Cython` and `GCC`. Add `libdbs_<py>_<arch>.so`.
- [DBS] New UI composed of 8 channels realtime & single channel frequency display
- [Record] Integrate recording interface to DBS HTML.
- [Streaming] Add function `send_message_streaming` for communicating from other process.

## 0.1.5 [2019-5-01]
- [DBS] Hide source code of data-processing algorithms with pre-compiled binary files.
- [Apps] Sub-app `WiFi` implemented as a web-based client of freedesktop NetworkManager, working on wireless card `AP6212` at STA+AP mode.
- [Apps] Support updating OTA (On-The-Air) by simple `git pull`.
- [WebUI] Upgrade `jQuery` to v3.4.1 and `Bootstrap` to v4.3.
- [WebUI] Immegrate from `glyphicon` to `font-awesome`.

## 0.1.4 [2019-3-23]
- [Doc] Add basic documentations.
- [Apps] Rename and move lots of DisplaySPI datafiles.
- [Apps] Add project `Speller`.
- [EmBCI] Add `PytestRunner` and func `test` in each module for runtime testing.
- [ESP32] New feature: turn into sleep mode to reduce power consumption.

## 0.1.3 [2019-3-11]
- [DBS] Change from `ESP32SPIReader` to `PylslReader` in embci.apps.DBS.
- [DBS] Save userinfo as cache to client(browser) for accessing report PDF.
- [Apps] Make data-streaming an individual task @ `embci.apps.streaming`.
- [Utils] Change `find_wifi_hotspots` output to type of `AttributeDict`.
- [WebUI] Move GeventWebsocketServer `log` to `logger` to log requests properly.


run_me1.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Thu Apr 26 19:27:13 2018

@author: hank
"""
# built-in
from __future__ import print_function
import os
import time
import threading
import subprocess
from functools import partial

# requiremnts.txt: necessary: numpy, scipy, pillow, reportlab, gpio4
# requiremnts.txt: optional: ipython
try:
    import IPython
    _NO_IPYTHON_ = False
except:
    _NO_IPYTHON_ = True
import numpy as np
from scipy import signal
from PIL import Image, ImageDraw
from reportlab.pdfbase import ttfonts, pdfmetrics
from reportlab.pdfgen import canvas

# from ./
from embci.common import check_input, time_stamp, mkuserdir
from embci.preprocess import Signal_Info
from embci.visualization import SPI_Screen_GUI as Screen_GUI
from embci.io import ESP32_SPI_reader as Reader, Socket_TCP_server


__dir__ = os.path.dirname(os.path.abspath(__file__))
__file__ = os.path.basename(__file__)


ILI9341_BLUE        = [0x00, 0x1F]  #   0   0 255
ILI9341_GREEN       = [0x07, 0xE0]  #   0 255   0
ILI9341_CYAN        = [0x07, 0xFF]  #   0 255 255
ILI9341_RED         = [0xF8, 0x00]  # 255   0   0
ILI9341_MAGENTA     = [0xF8, 0x1F]  # 255   0 255
ILI9341_YELLOW      = [0xFF, 0xE0]  # 255 255   0
ILI9341_WHITE       = [0xFF, 0xFF]  # 255 255 255
ILI9341_PURPLE      = [0x41, 0x2B]  # 128   0 128
ILI9341_ORANGE      = [0xFD, 0xC0]  # 255 160  10
ILI9341_GREY        = [0xCE, 0x59]  # 200 200 200

RGB_BLUE            = (  0,   0, 255)
RGB_GREEN           = (  0, 255,   0)
RGB_CYAN            = (  0, 255, 255)
RGB_RED             = (255,   0,   0)
RGB_MAGENTA         = (255,   0, 255)
RGB_YELLOW          = (255, 255,   0)
RGB_WHITE           = (255, 255, 255)
RGB_PURPLE          = (128,   0, 128)
RGB_ORANGE          = (255, 160,  10)
RGB_GREY            = (128, 128, 128)

ILI9341_Rainbow = [
    ILI9341_BLUE, ILI9341_YELLOW, ILI9341_MAGENTA, ILI9341_CYAN,
    ILI9341_GREEN, ILI9341_RED, ILI9341_PURPLE, ILI9341_ORANGE, ILI9341_GREY]
RGB_Rainbow = [
    RGB_BLUE, RGB_YELLOW, RGB_MAGENTA, RGB_CYAN,
    RGB_GREEN, RGB_RED, RGB_PURPLE, RGB_ORANGE, RGB_GREY]


def shutdown(*a, **k):
    for flag in flag_list:
        flag[1].set()
    s.close()
    reader.close()
    os.system('shutdown now')


def exit_program(*a, **k):
    for flag in flag_list:
        flag[1].set()
    s.empty_widget()
    s1 = 'Debug Mode'
    w, h = s.getsize(s1, size=25)
    s.draw_text((s.width-w)/2, (s.height-h)/2, s1, 'red')
    time.sleep(2)
    s.close()
    server.close()
    reader.close()


def reboot(*a, **k):
    for flag in flag_list:
        flag[1].set()
    s.close()
    reader.close()
    os.system('reboot')


@mkuserdir
def generate_pdf(username=u'三到四字', gender=u'男', id=0, age=20,
                 fontname='Mono', fontpath='files/fonts/yahei_mono.ttf',
                 **k):
    if fontname not in pdfmetrics.getRegisteredFontNames():
        font = ttfonts.TTFont(fontname, fontpath)
        pdfmetrics.registerFont(font)
    pdfname = os.path.join(__dir__, 'data', username,
                           k.get('filename', time_stamp() + '.pdf'))
    c = canvas.Canvas(pdfname, bottomup=0)
    c.setFont(fontname, 30)
    c.drawString(65, 80, u'天坛医院DBS术后调控肌电报告单')
    c.setFontSize(20)
    c.line(30, 120, 580, 120)
    c.drawString(35, 150,
                 (u'姓名:  {}  '
                  u'性别:  {}  '
                  u'年龄:  {:3d}  '
                  u'病号ID:  {:5d}  ').format(username, gender, age, id))
    c.line(30, 165, 580, 165)
    c.line(30, 710, 580, 710)
    c.drawString(35, 740, u'改善率    震颤： 80%    僵直： 80%    运动： 80%')
    c.line(30, 755, 580, 755)
    c.drawImage('aaa1.png', 35, 190)
    c.drawImage('./aaa1.png', 35, 450)
    c.setFontSize(24)
    c.drawString(360, 250, u'术前')
    c.drawString(360, 510, u'术后')
    c.setFontSize(18)
    c.drawString(380, 290, u'震颤： 12.345Hz')
    c.drawString(380, 320, u'僵直： 123.45')
    c.drawString(380, 350, u'运动： 10.0s')
    c.drawString(380, 550, u'震颤： 12.345Hz')
    c.drawString(380, 580, u'僵直： 123.45')
    c.drawString(380, 610, u'运动： 10.0s')
    c.drawString(35, 795, u'医师签字：                     Powered by Cheitech')
    c.save()
    print('pdf %s saved!' % pdfname)


def update(*a, **k):
    try:
        output = subprocess.check_output('git pull', shell=True,
                                         stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError as e:
        print('Update failed!' + str(e))
        s.freeze_frame()
        # TODO: draw element to show update failed
        s.recover_frame()
    else:
        # TODO: draw element to show update done
        print('Update success!\n' + output + '\nRebooting now...')
        reboot()


scale_list = {'a': [100, 200, 500,
                    1000, 2000, 5000,
                    10000, 20000, 50000,
                    100000, 200000, 500000,
                    1000000, 2000000, 5000000,
                    10000000],
              'i': 2}

channel_range = {'r': (0, 7), 'n': 0, 'step': 1}


page_list = {'a': ['./files/layouts/layout-DBS-page%d.pcl' % i
                   for i in range(6)],
             'i': 0}


def range_callback(r, operate, prev=None, after=None, *a, **k):
    if prev is not None:
        prev(*a, **k)
    if operate == 'plus':
        r['n'] += r['step']
        if r['n'] > r['r'][1]:
            r['n'] = r['r'][0]
    elif operate == 'minus':
        r['n'] -= r['step']
        if r['n'] < r['r'][0]:
            r['n'] = r['r'][1]
    else:
        return
    if after is not None:
        after(*a, **k)


def list_callback(l, operate, prev=None, after=None, *a, **k):
    if prev is not None:
        prev(*a, **k)
    if operate == 'next':
        l['i'] += 1
    elif operate == 'prev':
        l['i'] -= 1
    else:
        return
    l['i'] %= len(l['a'])
    if after is not None:
        after(*a, **k)


#                    flag_pause         flag_close
flag_list = [(threading.Event(), threading.Event()),  # page0
             (threading.Event(), threading.Event()),  # page1
             (threading.Event(), threading.Event()),  # page2
             (threading.Event(), threading.Event()),  # page3
             (threading.Event(), threading.Event()),  # page4
             (threading.Event(), threading.Event())]  # page5


def change_page(*a, **k):
    time.sleep(0.5)
    page_num = page_list['i']
    s.load_layout(page_list['a'][page_num], extend=False)
    for id in callback_list[page_num]:
        s.widget['button', id]['callback'] = callback_list[page_num][id]
    flag_list[page_num][0].set()
    flag_list[page_num][1].clear()
    func = globals().get('page%d_daemon' % page_num)
    if func is not None:
        threading.Thread(target=func, args=flag_list[page_num]).start()


def change_channel(*a, **k):
    s.widget['text', 15]['s'] = 'CH%d' % channel_range['n']
    s.render('text', 15)


def change_scale(*a, **k):
    scale = scale_list['a'][scale_list['i']]
    exp = int(np.log10(scale))
    s.widget['text', 16]['s'] = '%de%d' % (scale/10**exp, exp)
    s.render('text', 16)


test_dict = dict.fromkeys([(1, i) for i in range(2, 10)] +
                          [(4, i) for i in range(2,  8)])


def reverse_status(*a, **k):
    name = (page_list['i'], k['bt']['id'])
    test_dict[name] = not test_dict[name]


prev = partial(list_callback, l=page_list, operate='prev', after=change_page,
               prev=lambda *a, **k: flag_list[page_list['i']][1].set())
next = partial(list_callback, l=page_list, operate='next', after=change_page,
               prev=lambda *a, **k: flag_list[page_list['i']][1].set())


callback_list = [
    # page0
    # id of button: callback function
    {0: shutdown, 1: next, 2: update, 3: exit_program},
    # page1
    {0: prev, 1: next,
     2: reverse_status, 3: reverse_status,
     4: reverse_status, 5: reverse_status,
     6: reverse_status, 7: reverse_status,
     8: reverse_status, 9: reverse_status},
    # page2
    {0: prev, 1: next,
     2: partial(range_callback, r=channel_range,
                operate='minus', after=change_channel),
     3: partial(range_callback, r=channel_range,
                operate='plus', after=change_channel),
     4: partial(list_callback, l=scale_list,
                operate='next', after=change_scale),
     5: partial(list_callback, l=scale_list,
                operate='prev', after=change_scale)},
    # page3
    {0: prev, 1: next},
    # page4
    {0: prev, 1: next,
     2: reverse_status, 3: reverse_status, 4: reverse_status,
     5: reverse_status, 6: reverse_status, 7: reverse_status},
    # page5
    {0: prev, 1: next, 2: generate_pdf}]


def page1_daemon(flag_pause, flag_close, fps=0.8, thres=0):
    print('turn to page1')
    img_red = Image.open('./files/icons/4@300x-8.png').resize((21, 21))
    img_red = np.array(img_red.convert('RGBA'))
    img_green = Image.open('./files/icons/5@300x-8.png').resize((21, 21))
    img_green = np.array(img_green.convert('RGBA'))
    reader._esp.do_measure_impedance = True
    reader.data_channel
    last_time = time.time()
    last_status = [False] * 8
    while not flag_close.isSet():
        while (time.time() - last_time) < 1.0/fps:
            time.sleep(0.05)
        last_time = time.time()
        flag_pause.wait()
        data = reader.data_channel
        for i in np.arange(8):
            if test_dict[(1, i+2)]:
                s.widget['button', i+2]['s'] = '{:.1e}'.format(data[i])
                s.render('button', i+2)
                if data[i] > thres and not last_status[i]:
                    s.widget['img', i+5]['img'] = img_green
                    s.render('img', i+5)
                    last_status[i] = True
                elif data[i] < thres and last_status[i]:
                    s.widget['img', i+5]['img'] = img_red
                    s.render('img', i+5)
                    last_status[i] = False
    reader._esp.do_measure_impedance = False
    print('leave page1')


def page2_daemon(flag_pause, flag_close, step=1, low=1.5, high=80.0,
                 area=[39, 45, 290, 179], center=112):
    print('turn to page2')
    change_scale()
    line = np.ones((2, area[2]), np.uint16) * center
    data = reader.data_frame
    si.bandpass(data, low, high, register=True)
    si.notch(data, register=True)
    while not flag_close.isSet():
        flag_pause.wait()
        ch = channel_range['n']
        scale = scale_list['a'][scale_list['i']]
        c = ILI9341_Rainbow[ch]
        s._ili.draw_rectf(area[0], area[1],
                          area[0]+step*3, area[3], ILI9341_WHITE)
        s.widget['text', 17]['s'] = u'%5.1fs\u2191' % \
            (time.time() - reader._start_time)
        s.render('text', 17)
        for x in np.arange(area[0], area[2], step, dtype=int):
            if flag_close.isSet():
                break
            d = reader.data_channel
            server.send(d)
            data = d[ch]
            filtered = si.bandpass_realtime(si.notch_realtime(data))
            data = center - np.array([data, filtered]) * scale
            line[:, x] = np.uint16(np.clip(data, area[1], area[3]))
            yraw, yflt = line[:, x-step:x+1]
            _x = min((x + step*4), area[2])
            s._ili.draw_line(_x, area[1], _x, area[3], ILI9341_WHITE)
            if yraw[0] == yraw[1]:
                s._ili.draw_point(x, yraw[0], ILI9341_GREY)
            else:
                s._ili.draw_line(x, yraw.min(), x, yraw.max(), ILI9341_GREY)
            if yflt[0] == yflt[1]:
                s._ili.draw_point(x, yflt[0], c)
            else:
                s._ili.draw_line(x, yflt.min(), x, yflt.max(), c)
    print('leave page2')


def page3_daemon(flag_pause, flag_close, fps=0.6, area=[26, 56, 153, 183]):
    print('turn to page3')
    last_time = time.time()
    x = np.linspace(0, reader.sample_time, reader.window_size)
    sin_sig = 1e-3 * np.sin(2*np.pi*32*x).reshape(1, -1)
    x = np.arange(127).reshape(1, -1)
    blank = np.zeros((127, 127, 4), np.uint8)
    while not flag_close.isSet():
        flag_pause.wait()
        if (time.time() - last_time) < 1.0/fps:
            continue
        last_time = time.time()
        ch = channel_range['n']
        c = RGB_Rainbow[ch]
        d = reader.data_frame
        server.send(d)
        d = si.notch(d[ch])
        s.widget['text', 23]['s'] = '%.2f' % movement_coef(d)
        s.render('text', 23)
        d = si.detrend(d)
        amp = si.fft(sin_sig + d, resolution=4)[1][:, :127]
        amp[0, -1] = amp[0, 0] = 0
        amp = np.concatenate((x, 127 * (1 - amp / amp.max()))).T
        img = Image.fromarray(blank)
        ImageDraw.Draw(img).polygon(map(tuple, amp), outline=c)
        s._ili.draw_rectf(*area, c=ILI9341_WHITE)
        s._ili.draw_img(area[0], area[1], np.uint8(img))
        s.widget['text', 21]['s'] = '%.2f' % tremor_coef(d)[0]
        s.render('text', 21)
        s.widget['text', 22]['s'] = '%.2f' % stiffness_coef(d)
        s.render('text', 22)
    print('leave page3')


def page4_daemon(flag_pause, flag_close, fps=0.8):
    print('turn to page4')
    start_time = [None, None]
    last_time = time.time()
    while not flag_close.isSet():
        flag_pause.wait()
        if (time.time() - last_time) < 1.0/fps:
            continue
        last_time = time.time()
        ch = channel_range['n']
        d = si.notch(si.detrend(reader[ch]))
        for i in np.arange(2, 8):
            if test_dict[(4, i)]:
                if i % 3 == 2:
                    freq = tremor_coef(d)[0]
                    s.widget['button', i]['s'] = '  %5.1f ' % freq
                    s.render('button', i)
                elif i % 3 == 0:
                    stiff = stiffness_coef(d)
                    s.widget['button', i]['s'] = '  %5.1f ' % stiff
                    s.render('button', i)
                else:
                    if not start_time[(i-2)/3]:
                        start_time[(i-2)/3] = time.time()
                    t = time.time() - start_time[(i-2)/3]
                    s.widget['button', i]['s'] = '  %4.1fs ' % t
                    s.render('button', i)
            elif i % 3 == 1 and start_time[(i-2)/3]:
                start_time[(i-2)/3] = None
        for i in np.arange(21, 24):
            if not test_dict[(4, i-19)] and not test_dict[(4, i-16)]:
                continue
            b, a = s.widget['button', i-19, 's'], s.widget['button', i-16, 's']
            if 'test' in b or 'test' in a:
                continue
            try:
                b, a = float(b[:-2]), float(a[:-2])
                s.widget['text', i]['s'] = '%.2d%%' % (abs(b-a) / b * 100)
                s.render('text', i)
            except:
                pass
    print('leave page4')


def tremor_coef(data, ch=0, distance=25):
    data = si.smooth(si.envelop(data), 15)[0]
    data[data < data.max() / 4] = 0
    peaks, heights = signal.find_peaks(data, 0, distance=si.sample_rate/10)
    peaks = np.concatenate(([0], peaks))
    return (si.sample_rate / (np.average(np.diff(peaks)) + 1),
            1000 * np.average(heights['peak_heights']))


def stiffness_coef(data, ch=0):
    b, a = signal.butter(4, 10.0/si.sample_rate, btype='lowpass')
    return 1000 * si.rms(signal.lfilter(b, a, data, -1))


def movement_coef(data, ch=0):
    data = si.smooth(si.envelop(data), 10)[0]
    return 1000 * np.average(data)


if __name__ == '__main__':
    username = 'test'
    try:
        print('username: ' + username)
    except NameError:
        username = check_input('Hi! Please offer your username: ', answer={})

    # reset_esp()

    reader = Reader(sample_rate=500, sample_time=2, n_channel=8)
    reader.start()
    server = Socket_TCP_server()
    server.start()
    si = Signal_Info(500)
    s = Screen_GUI()
    change_page()
    if _NO_IPYTHON_:
        s.start_touch_screen('/dev/ttyS1')
    else:
        s.start_touch_screen('/dev/ttyS1', block=False)
        IPython.embed()
